teps/draft-20200716-task-results-in-final-tasks.md

```diff
@@ -0,0 +1,304 @@
+
+Second goal of this proposal is to determine what happens to the final `Task` when referenced
+task results are not initialized because either dependent task failed or dependent task skipped
+due to its own condition or if its parent was skipped.
+
+### Non-Goals
+
+This proposal is not enabling final `Task` to receive `Results` from other final `Tasks`.
+
+This proposal is also not enabling final `Tasks` to produce `Task` `Results`.
+
+## Proposal
+
+Allow a finally `Task` to get a value from a non-finally `Task` via a `Task` `Result`.
+If the `Task` producing the value fails, the finally `Task` that requires that value will be marked as failed.
+That final `Task` will be `attempted` i.e. evaluated by the `Pipeline Controller` for execution,
+but the validation will fail before the `TaskRun` is created since the required `Result` is not available.
+However if the `Task` being attempted has a default value for that parameter, it can still run,
+using the default when result is not available.
+
+### User Stories
+
+A `Pipeline` author can send the name of the project to finally task after non-finally task creates/acquires
+resources for that project. The finally `Task` can then clean up those project resources regardless of the
+failure of the rest of the `Tasks` in the Pipeline.
+
+## Design Details
+
+`Tasks` can produce one or more execution `Results` under `/tekton/results/` by specifying `results`
+in the `Task` specification:
+
+```yaml
+spec:
+  results:
+    - name: leased-resource
+      description: The name of the leased resource
+  steps:
+```
+
+This task produces a result file at `/tekton/results/leased-resource`.
+
+Any other `Pipeline` `Tasks` can use these `Results` as parameter values through variable substitution:
+
+```yaml
+spec:
+  tasks:
+    - name: boskos-release
+      params:
+        - name: leased-resource
+          value: $(tasks.boskos-acquire.results.leased-resource)
+```
+
+No need to specify explicit runAfter here, `Tekton` infers the execution order by setting `boskos-acquire`
+as a dependency so that the task emitting the referenced results executes before the task that consumes them.
+
+Similar to non-final tasks, final task use `Results` through variable substitution:
+
+```yaml
+spec:
+ tasks:
+   - name: boskos-acquire
+     taskRef:
+       name: boskos-acquire
+   - name: do-stuff-with-resource
+     taskRef:
+       name: do-stuff-with-resource
+     params:
+       - name: resource-name
+         value:  $(tasks.boskos-acquire.results.leased-resource)
+ finally:
+   - name: boskos-release
+     taskRef:
+       name: boskos-release
+     params:
+       - name: leased-resource
+         value:  $(tasks.boskos-acquire.results.leased-resource)
+```
+
+**Note:** Today final tasks are executed all in parallel and can not depend on any other `Task` in the `Pipeline`.
+With adding support for `Task` `Results`, we are introducing dependencies to finally `Tasks` but that
```

Comment:
I am not sure I understand this (nor if it should be there). What we are saying is "now finally tasks depends on tasks" ? that was implicitly the case already (from the user perpective) as the finally task will wait for all the "normal" task to complete (success or failure).

Reply:
> What we are saying is "now finally tasks depends on tasks" ? that was implicitly the case already (from the user perpective) as the finally task will wait for all the "normal" task to complete (success or failure).

Finally tasks did not have any implicit/explicit dependency on any other task. The scheduling algorithm sends final tasks to execute after all normal/DAG tasks are completed. Now, with task results, implicit dependency is being introduced. There is no impact on scheduling yet, final tasks still scheduled after all DAG. But worth calling out now and will easier to understand what happens when we add support for task results from other final tasks, `runAfter` in final tasks, `from` in final tasks, etc